Changes based on Cloudwatch Agent work

JamesTimms opened the pull request.

Commit:
Removing boto install

Should be managed by requirements.txt instead.

Commit:
Changed aws keys to force them to be specified

Having keys implicitly loaded from env from a role is a hidden requirement. Better to make it explicit and let the keys be provided from the playbook level.

Commit:
Updated log_groups with cleaner token code

Removed multiple variables for aws key and secret and added security_token for when there is no access key or secret set. These variables should be loaded from env or specified using aws_acces_key, aws_secret_key or security_token manually when needed.

Commit:
Updated main to not error out in normal conditions

Previously the 'Already Install' task would error out when cloudwatch agent wasn't install. Now it doesn't error out when installed or not installed and following tasks now use the return code instead of failure status.

JamesTimms commented on tasks/main.yml:
```diff
@@ -86,13 +87,6 @@
-      - boto3
-  delegate_to: localhost
-
 - name: Setup log groups
```
Boto3 should be installed before running this role. I can't see any reason to make the role install it, instead if boto3 isn't installed then the role should fail.

jeremypoulter replied:
True I guess

JamesTimms commented on tasks/main.yml:
```diff
@@ -45,7 +45,9 @@
     name: amazon-cloudwatch-agent
     state: present
-  ignore_errors: yes
+  failed_when: 
+    - not (already_inst.rc == 0)
+    - not (already_inst.rc == 126)
```
This was failing normally which makes it harder to tell if the role worked properly. If this task actually failed it would have caused problems.

Now, if `amazon-cloudwatch-agent` is installed this will succeed with `rc` of 0. If it's not installed then it will now succeed but with `rc` of 126.

jeremypoulter replied:
Good change although I have never been happy with this

JamesTimms commented on tasks/log_groups.yml:
```diff
@@ -20,8 +20,9 @@
       metric_value: "{{ filter.metric_transformation.metric_value }}"
-    aws_access_key: "{{ deploy_aws_access_key | default(aws_access_key | default(lookup('env','AWS_ACCESS_KEY_ID'))) }}"
-    aws_secret_key: "{{ deploy_aws_secret_key | default(aws_secret_key | default(lookup('env','AWS_SECRET_ACCESS_KEY'))) }}"
+    aws_access_key: "{{ aws_access_key | default(omit) }}"
+    aws_secret_key: "{{ aws_secret_key | default(omit) }}"
+    security_token: "{{ security_token }}" # Use token to allow for assumed roles
```
I can't see any reason to hard code so many variables into a role for the same value. Also, env lookups shouldn't be done inside a role but instead, there should be fixed variables that are passed to the role.

I've also added `security_token` which I think might be needed for assumed roles but I'm not sure. Might be able to remove `security_token` if it turns out assumable roles are fine using access key and secret.

JamesTimms replied:
I've changed these back to the module's expected variables.

jeremypoulter replied:
`security_token` is unlikely to be needed. any assumed role is likely to be set via env vars so should be picked up automatically.

As commented elsewhere the whole `{{ deploy_aws_access_key | default(aws_access_key | default(lookup('env','AWS_ACCESS_KEY_ID'))) }}"` is a mess and probably the defaults are better, but this is used in a bunch of places and needs removing across the board so better to do as a separate task

JamesTimms replied:
I'm happy to remove `security_token` as I was never sure if it was needed. There was one example where it was needed in a different playbook but I think my inventory or envs were the issue there.

I'll also remove the omit as without `security_token` we want this to fail if the key and secrets are missing :)

Commit:
Changing back variables to author's original

Noticed the original author used specific variables for AWS credentials and region, so I'm switching back to those instead of our specific variables.

jeremypoulter commented on tasks/log_groups.yml:
```diff
@@ -1,13 +1,12 @@
-    region: "{{ aws_region }}"
+    retention: "{{ item.log_retention|int }}"
+    tags: "{{ item.log_group_tags }}"
+    aws_access_key: "{{ aws_cwa_key_access | default(omit) }}"
```
Not sure this is a good move, `aws_cwa_key_access`/`aws_cwa_key_secret`, other than being a terrible name, are the runtime credentials, this change forces them to have more permissions that are only required during the setup phase.

JamesTimms replied:
I don't understand sorry. I've assumed these are only used in this role and not in `wg-ansible` at all. So, in this case, these variables become required by the role, so they can be set when the role is called or in wg-ansible inventories.

jeremypoulter replied:
See https://github.com/Webgains/ansible-role-amazon-cloudwatch-agent/blob/master/tasks/main.yml#L24:L37, that task sets up the credentials that the agent uses at run time, this task is run during instillation and need different permissions than the agent requires at run time. By linking the two together you are forcing the runtime credentials to have greater permissions than needed.

That being said this is a mess and needs to be removed, but probably better to do as a separate task as it will change a lot of places.

JamesTimms replied:
So with the current use case what will happen is the credential file created on the remote will use the currently assumed aws_key and aws_secret instead of some shared key or a specific user's credentials.

If run via the common role it'll use `host_aws_access_key_id` and `host_aws_secret_access_key`

JamesTimms commented:
I've just noticed that we're storing `aws_cwa_key_access` and `aws_cwa_key_secret` in `/root/.aws/credentials`. This is a security risk as each EC2 instance will have credentials to gain access to the console.

So either the user used here need to be locked down or we should remove them. As far as I remember only on-prem needs credentials and a user. So I'm not sure if this is always needed...

jeremypoulter commented:
Yes we should be using roles for EC2 instances, but baby steps, get it working first and iterate, and indeed the logging user should only be able to do just that.

JamesTimms commented:
So `src: "https://s3.amazonaws.com/amazoncloudwatch-agent/linux/amd64/latest/AmazonCloudWatchAgent.zip"` is the agent for on-prem then? I did notice it wasn't the same link from the guides...

I can setup a user in each env and then setup inventories for them... OR I could update this role to work with EC2 and on-prem. I think the later would take a bit longer but easier to do it now than in a few weeks.

jeremypoulter commented:
I do not quite understand, src: "https://s3.amazonaws.com/amazoncloudwatch-agent/linux/amd64/latest/AmazonCloudWatchAgent.zip" is used for both instances, I do not think there is a different client for different use cases.

For now I think using a logging user for each env is the easiest way forward, migrating to using instance roles is a much larger task and expands outside of just this one use case so should be tackled separately

JamesTimms commented:
https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/download-cloudwatch-agent-commandline.html
I think the CloudWatch agent is different for different OSes. I thought it was different for EC2 and on-prem as well but I must have misread.

JamesTimms commented:
Also https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/create-iam-roles-for-cloudwatch-agent-commandline.html

> If you're going to use the agent on Amazon EC2 instances, you must create an IAM role. If you're going to use the agent on on-premises servers, you must create an IAM user.

jeremypoulter commented:
Sure different OS have a different install package but that has nothing to do with EC2/on-prem. Not sure why the original author is using the zip and maybe a future improvement could be to use the deb/rpm

Commit:
Removing omit as no longer using security_token sometimes

Commit:
Updated default variables

Created new variables and remove unneeded code.

JamesTimms commented:
Ok I'm happy with these changes for now. I'll add a task to pick up on the security issue and lack of EC2 role.

JamesTimms commented:
@jeremypoulter , I get the below when there is no `security_token` field while assuming the fullaccess London dev role. As soon as I add the `security_token` field in it works every time. I've ruled out the dynamic inventory iam_arn read-only role but otherwise, I've no idea why this is failing.

The precedence means the key and secrets are tried first then the `security_token` so I assume it's failing with the key and secret then succeeding with the security token under the hood.

```
TASK [aws-cloudwatch-agent : Ensure log groups /webgains/dev/test exist] ***********************************
task path: /home/jtimms/wg-ansible/roles/aws-cloudwatch-agent/tasks/log_groups.yml:2
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: botocore.errorfactory.UnrecognizedClientException: An error occurred (UnrecognizedClientException) when calling the DescribeLogGroups operation: The security token included in the request is invalid.
fatal: [test-for-cloudwatch-3]: FAILED! => {
    "changed": false,
    "error": {
        "code": "UnrecognizedClientException",
        "message": "The security token included in the request is invalid."
    },
    "response_metadata": {
        "http_headers": {
            "connection": "close",
            "content-length": "107",
            "content-type": "application/x-amz-json-1.1",
            "date": "Wed, 23 Sep 2020 08:45:28 GMT",
            "x-amzn-requestid": "27bc7c28-f407-4159-9ceb-d5168af1654a"
        },
        "http_status_code": 400,
        "request_id": "27bc7c28-f407-4159-9ceb-d5168af1654a",
        "retry_attempts": 0
    }
}

MSG:

Unable to describe log group /webgains/dev/test: An error occurred (UnrecognizedClientException) when calling the DescribeLogGroups operation: The security token included in the request is invalid.
```

```
---
- name: Ensure log groups {{ item.log_group_name }} exist 
  cloudwatchlogs_log_group:
    log_group_name: "{{ item.log_group_name }}"
    retention: "{{ item.log_retention|int }}"
    tags: "{{ item.log_group_tags }}"
    aws_access_key: "{{ aws_cwa_key_access }}"
    aws_secret_key: "{{ aws_cwa_key_secret }}"
    region: "{{ aws_cwa_region }}"
  delegate_to: localhost
```

JamesTimms commented:
I've found out the implicit env values aren't being used by the `cloudwatchlogs_log_group` module due to `become: yes`. What happens is the localhost is used to run the task but then Ansible sudo su's to root which doesn't have the env variables set. So when this is working correctly it'll implicitly use the `security_key` variable but as `root@localhost` is the effective user this fails.

Note this shows that `security_key` is in fact required while using an assumed role. Might mean we have a bunch of issues in other playbooks!

Commit:
Adding security_token back and updating README

Adding security_token back in as it's needed for assumed roles. Also forcing the cloudwatch tasks run locally to be run without assuming root so the module can implicitly use env variables.

Found at that while using an assumed role the cloudwatch modules implicitly use the security_token from the env variables however due to the role requiring sudo via  it turned out that the env variables weren't there (for root) and so the module failed unless the security_token was set explicitly.

jeremypoulter commented:
I would kind of think it just highlights that we are doing too much as root ;-) I think you can solve this by adding `become: no` to the task and probably something to watch out for in any task delegated to `localhost`

JamesTimms commented:
Yeah really become should be no on the playbook level and yes on specific tasks. Sadly this role needs `become: yes` due to the credentials being set on the root level. Might update later to use a none root user (which is recommended) and then use become on the task but set the become_user to the cloudwatch agent specific user.

jeremypoulter commented:
I mean add it to the task, I believe an explicit `become: no` on the task will override anything at a higher level.

JamesTimms commented:
Already have :)
https://github.com/Webgains/ansible-role-amazon-cloudwatch-agent/pull/2/commits/a7513ae41c62e7c9d5f78d4f3e60761b0f50f2e9

Commit:
Updateing the Agent config to support assumable roles and none root users

Can now use an assumable role to pass logs to and a none root local user. If these new variables aren't specified then the config is omitted from the config file, which results in no change from before. Still needs testing.

Commit:
Adding check to exclude local credentials

Local credentials are only needed on-prem. EC2 instances should only need an attached role on the EC2 instance with the correct permissions which should be done outside of this role.

Commit:
Updating log group AWS credentials

Changing the authentication credentials back to be separate from the CloudWatch Agent local credentials used for on-prem setups.

JamesTimms commented on tasks/log_groups.yml:
```diff
@@ -1,14 +1,15 @@
-    region: "{{ aws_region }}"
+    retention: "{{ item.log_retention|int }}"
+    tags: "{{ item.log_group_tags }}"
+    aws_access_key: "{{ deploy_aws_access_key | default(lookup('env','AWS_ACCESS_KEY_ID')) }}"
+    aws_secret_key: "{{ deploy_aws_secret_key | default(lookup('env','AWS_ACCESS_KEY_ID')) }}"
+    security_token: "{{ security_token | default(omit) }}"
```
I've changed this back to use `deploy_aws_access_key` and `deploy_aws_secret_key` by default but these should be treated as unique and separate from wg-ansible credentials and so set explicitly in wg-ansible where it makes sense. That way if later we change the name of that variable in wg-ansible we won't break this role!

jeremypoulter replied:
I think this is ok. The only other thing you may need to change (other than the other mass of similar bits of code) is https://github.com/Webgains/wg-ansible/blob/fe66982103914f9a096f3fe7a1752072f58a204f/Vagrantfile#L172 that should be changed to set `deploy_aws_access_key` etc. 

However I still think it would just be better to leave as the original for now as that is 'working'. Then tackle the issue as a whole. Ultimately these lines just need removing with the assumption that the AWS SDK is configured with the correct credentials outside of Ansible.

JamesTimms replied:
The original was no working sadly. This change had to be made to support assumable roles.

jeremypoulter replied:
Adding the `security_token` is what was wrong with role as you need all three vars, but the `aws_access_key` and `aws_secret_key` should be fine as they where, and as commented elsewhere `security_token` should follow a similar model until removed

JamesTimms replied:
So are you saying the `security_token` should be used while assuming a role and while only using a user?

JamesTimms commented on tasks/log_groups.yml:
```diff
@@ -20,11 +21,13 @@
-    aws_access_key: "{{ deploy_aws_access_key | default(aws_access_key | default(lookup('env','AWS_ACCESS_KEY_ID'))) }}"
-    aws_secret_key: "{{ deploy_aws_secret_key | default(aws_secret_key | default(lookup('env','AWS_SECRET_ACCESS_KEY'))) }}"
-    region: "{{ aws_region }}"
+    aws_access_key: "{{ deploy_aws_access_key | default(lookup('env','AWS_ACCESS_KEY_ID')) }}"
+    aws_secret_key: "{{ deploy_aws_secret_key | default(lookup('env','AWS_ACCESS_KEY_ID')) }}"
+    security_token: "{{ security_token | default(omit) }}"
```
Same as the previous comment, https://github.com/Webgains/ansible-role-amazon-cloudwatch-agent/pull/2/files#r494202904

jeremypoulter replied:
Ditto ;-)

Commit:
Merge branch 'refresh' into update_config

Commit:
Adding debug option for CloudWatch Agent

Added an optional debug flag which turns on debug messages.

Commit:
Fixing typos and adding endpoint support

Fixing aws auth cred typo, Agent config debug typo and adding optional endpoint_override in agent config.

jeremypoulter commented on tasks/log_groups.yml:
```diff
@@ -1,14 +1,15 @@
+    tags: "{{ item.log_group_tags }}"
+    aws_access_key: "{{ deploy_aws_access_key | default(lookup('env','AWS_ACCESS_KEY_ID')) }}"
+    aws_secret_key: "{{ deploy_aws_secret_key | default(lookup('env','AWS_ACCESS_KEY_ID')) }}"
+    security_token: "{{ security_token | default(omit) }}"
```
Whatever we end up doing `security_token` should follow a similar pattern as well, but also again this is a wider problem than just this role, everywhere that uses `deploy_aws_access_key` will need the same change...

JamesTimms replied:
So just to be clear, the global `deploy_aws_acces_key` and `deploy_aws_secret_key` has to go. It causes this to fail when using an assumed role. This only effects London-architecture-based inventories for now so I'm only removing it from those, but a global user or auth token is a security risk. Easier to just use your own users and assume roles that have permissions to perform the action you need to perform.

jeremypoulter replied:
For sure it has to go, but you will have this issue in may places so better to do it as an explicit task rather than bit at a time so it is a just how it works, avoids doing one thing for one inventory and another, and/or breaking other use cases.

Commit:
Updated credentials and agent template

Added check to remove aws credentials when using EC2 over on-prem. Fixed typos in cloudwatch agent config template. Added default git cloudwatch log file.

Commit:
Reverting aws creds back to original variables

Variables need a review but now isn't the time, so I'm reverting them back to the original variables to keep things working across all inventories.

Commit:
Merge pull request #3 from Webgains/update_config

Updated Agent config

JamesTimms commented:
I've merged in some changes from PR #3 here to make it easier to review. Some changes made on PR #3 were bug fixes where some were functional changes. Some functional changes were needed others are nice to haves.

jeremypoulter commented on tasks/main.yml:
```diff
@@ -1,32 +1,48 @@
+- name: Add AWS CWA profile to /{{ aws_cwa_local_user }}/.aws/credentials
   blockinfile:
-    path: '/root/.aws/credentials'
+    path: "/{{ aws_cwa_local_user }}/.aws/credentials"
```
That is only valid for `root` other users would be `/home/{{ aws_cwa_local_user }}`

Commit:
Fixing use case for none root users

The previous example only worked for root as 'home/' wasn't being added to the path.

Commit:
Removed debug code

Commit:
Merge remote-tracking branch 'origin/refresh' into refresh

Commit:
Commented on retention usage

If log groups have several log streams with differing retention periods, the last log stream's retention period will override the others.

JamesTimms commented:
I've fixed the root user versus none root user issue (but haven't tested properly). I've also added a comment about a bug with retention and log streams. I'm not going to address this as it's something that can be fixed with correct variables in each log group. Might be worth adding some sort of warning or error, as if the logs are set to 7 days and one stream in the group is set to 3 days, the logs from days 4-7 will be deleted while re-running the playbook.

The below will delete logs from days 4-7 and finish with a retention policy of 7 days as the last entry is set to 7 days
```
[
  {
  ...
  retention: 30
  ...
  },
  {
  ...
  retention: 3
  ...
  },
  {
  ...
  retention: 7
  ...
  },
  {
  ...
  retention: 7
  ...
  },
]
```

JamesTimms commented:
This should be is minus the bug I've mentioned I'm still debugging.

Commit:
Fixing issues with root path

It's '/root' not '/'

Commit:
Fixing issue with endpoints

Turns out there is a logging endpoint and a monitoring endpoint. Previous was using the logging endpoint for both logging and monitoring which caused custom metrics to be rejceted.

jeremypoulter commented on defaults/main.yml:
```diff
@@ -4,3 +4,11 @@ aws_cwa_disk_monitor_paths:
 aws_cwa_cfgs: generic-amazon-cloudwatch-agent.json.j2
+aws_cwa_url: "https://s3.amazonaws.com/amazoncloudwatch-agent/linux/amd64/latest/AmazonCloudWatchAgent.zip"
+aws_cwa_local_user: root # The user the CloudWatch Agent runs as.
+aws_cwa_user_path: "/"
```
Should this not be
```suggestion
aws_cwa_user_path: "/root"
```

JamesTimms replied:
Yup, already fixed but forgot to push last night's changes 🙄

jeremypoulter commented on tasks/main.yml:
```diff
@@ -1,32 +1,55 @@
+  set_fact:
+    aws_cwa_user_path: "/home/{{ aws_cwa_local_user }}"
+  when:
+    - aws_cwa_local_user != "root"
+    - aws_cwa_local_user == "/" # When not default, assume user has overridden.
```
Use `and` or `or` plus is that second condition correct? should it be `aws_cwa_user_path` ?

JamesTimms replied:
This is wrong but I thought I fixed this already. Maybe I forgot to push???

JamesTimms replied:
Yes, I did... Sorry.

jeremypoulter commented on tasks/main.yml:
```diff
@@ -1,32 +1,55 @@
       output = json
       region = {{aws_cwa_region|default('us-east-1')}}
   notify: restart amazon-cloudwatch-agent
+  when: aws_cwa_on_prem
```
I do not think this needs to be an explicit var to setup for 'on-prem' (needs renaming if we keep) can just check for `aws_cwa_key_access` and `aws_cwa_key_secret` being set

JamesTimms replied:
I think that's going to hide a bunch of issues, especially as existing inventories ALWAYS have those values defined. Especially with the change, you're asking for on the other PR.

At best they'll be undefined or empty strings, but more likely they'll be set to AWS Ireland deploy access key and secret when setting up Pulsant due to the global default and vault files.

JamesTimms replied:
We could do;
`aws_cwa_on_prem: "{{ aws_cwa_key_access is not defined or aws_cwa_key_access | length < 0 }}"`

jeremypoulter replied:
That is ok, would clear it up. I do not think  `aws_cwa_key_access` is always defined if you change as I suggest:

```
    aws_cwa_key_access: "{{ host_aws_access_key_id | default(omit) }}"
    aws_cwa_key_secret: "{{ host_aws_secret_access_key | default(omit) }}"
```

`host_aws_access_key_id` is currently only defined in the `aws_scratch` and `feature_dev` inventories, which can probably be removed as these changes are rolled out as those use cases will use the instance role

jeremypoulter commented on tasks/main.yml:
```diff
@@ -1,32 +1,55 @@
+    aws_cwa_user_path: "/home/{{ aws_cwa_local_user }}"
+  when:
+    - aws_cwa_local_user != "root"
+    - aws_cwa_user_path == "/root" # When not default, assume user has overridden.
```
Please use `and`, I certainly can never remember if an array is `and` or `or`

JamesTimms replied:
Arrays are `and`. I think this is much clearer than long lines of `and` and `or` but I'll switch back if you'd like.

jeremypoulter reviewed (commented):
Re setting the `security_token` are the env vars preserved now the task is run with `become: no` and can be removed?

Commit:
Updating on-prem default and formatting conditional

Updating on-prem default to a conditional based on aws keys and changing array based and conditional to  conditional as requested.

JamesTimms commented:
I'd rather keep `security_token` as explicit for now to avoid hidden bugs and confusion. We need to decide how to go about AWS access variables in general. I don't like the idea of having modules implicitly use the env variable as it makes things opaque.

Also, we need to consider how this will work with the Ireland-based architecture and the London-based architecture. In London we're using assumable roles which need to use `security_token`. Currently, roles have to be assumed manually before running playbooks but I'd like to change that so roles are assumed based on variables in the inventory so we can fix which roles are used to run certain tasks. This way it's harder to assume the wrong role and means less mental effort spent trying to figure out which role you need to assume.

If we remove this and only use env variables implicitly, then we can't automatically assume roles set in inventories.

jeremypoulter commented:
>I'd rather keep security_token as explicit for now to avoid hidden bugs and confusion. We need to decide how to go about AWS access variables in general. I don't like the idea of having modules implicitly use the env variable as it makes things opaque.

In that case please change to

```
security_token: "{{ deploy_security_token | default(omit) }}"
```

so it is named similar to the matching access key/secret.

> Currently, roles have to be assumed manually before running playbooks

that is fine and is supported by just removing

> but I'd like to change that so roles are assumed based on variables in the inventory so we can fix which roles are used to run certain tasks. This way it's harder to assume the wrong role and means less mental effort spent trying to figure out which role you need to assume.

While this is true, that reasoning is also the reason we have this mess in the first place, (all be it with a user rather than a role). I really think setting the AWS credential/roles/whatever is a problem to solve outside of Ansible (but not in this PR ;-) )

> If we remove this and only use env variables implicitly, then we can't automatically assume roles set in inventories.

It can change ;-) lets consider the current use case not a hypothetical future use case

Commit:
Changing security token var name

Changing from security_token to deploy_security_token for consistency.

JamesTimms commented:
> >but I'd like to change that so roles are assumed based on variables in the inventory so we can fix which roles are used to run certain tasks. This way it's harder to assume the wrong role and means less mental effort spent trying to figure out which role you need to assume.

>While this is true, that reasoning is also the reason we have this mess in the first place, (all be it with a user rather than a role). I really think setting the AWS credential/roles/whatever is a problem to solve outside of Ansible (but not in this PR ;-) )

You're right partly right, it is a problem to solve outside of this PR but I think this is different from how AWS creds are set now. All this should amount to is a single variable `aws_iam_role_arn_to_assume` which can be overridden via `--extra-vars` or similar. This variable would be set at the host or service level.

Commit:
Reverting Already Installed due to bug

The package module returns different values depending on the package manage it uses. For centos, yum returns 'rc'. For ubuntu, apt returns only stdout and stderr.

I tried to rewrite this using conditionals instead of returning to the forced failed task (which is bad code) but the conditonal statements became to complex and didn't make any sense. I think there's a better solution but it's going to take too long to figure it out so I'm returning to the old method that was working dispite forcing the task to fail.

JamesTimms commented:
Reverted `Already Install?` due to a bug (see last commit comments). Also found another bug that only effects Ireland if you're assuming a role while running the this role.

For some reason, even if you omit `security_token: "{{ deploy_security_token | default(omit) }}"` the module `cloudwatchlogs_log_group` uses your env variable for `security_token` anyway which causes an error as the security token used by static credentials (from vault) are not compatible with the security_token in your env vars. If you're not assuming a role then there is no issue.

The fix is to not assume a role. I'll have to update `wg-ansible` with my scrip from `Infrastructure` which allows you to stop assuming a role via `. ./iam-assume-role.sh --stop`.

JamesTimms commented:
@jeremypoulter , am I still good to merge this with the last commit?

JamesTimms commented:
For the interest of time, I'm going to assume this last change is ok with you. I've tested it and it works so I'm happy.
